### PR TITLE
Refactor creating an application to be a command

### PR DESCRIPTION
**Pull request by luontola**
Followup from #1519/#832

# Definition of Done / Review checklist

## Reviewability
- [x] link to issue
- [x] note if PR is on top of other PR
- [x] note if related change in rems-deploy repo
- [x] consider adding screenshots for ease of review

## API
- [x] API is documented and shows up in Swagger UI
- [x] API is backwards compatible or completely new
- [x] Events are backwards compatible

## Documentation
- [x] update changelog if necessary
- [x] add or update docstrings for namespaces and functions
- [x] components are added to guide page
- [x] documentation _at least_ for config options (i.e. docs folder)
- [x] ADR for major architectural decisions or experiments

## Different installations
- [x] new configuration options added to rems-deploy repository
- [x] instance specific translations (i.e. LBR kielivara)

## Testing
- [x] complex logic is unit tested
- [x] valuable features are integration / browser / acceptance tested automatically

## Accessibility
- [x] all icons have the aria-label attribute
- [x] all fields have a label
- [x] errors are linked to fields with aria-describedby
- [x] contrast is checked
- [x] conscious decision about where to move focus after an action

## Follow-up
- [x] new tasks are created for pending or remaining tasks
- [x] no critical TODOs left to implement

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,11 @@ have notable changes.
 
 Changes since v2.4
 
+### Breaking changes
+- `:application/external-id` has been made a non-optional field.
```
Comment (opqdonut) [inline CHANGELOG.md]: ... in the API?

File: src/clj/rems/application/commands.clj [hunk 2]
```diff
@@ -256,6 +258,58 @@
...
+                                   {:keys [allocate-application-ids! get-catalogue-item get-workflow]
+                                    :as injections}]
+  (assert (seq catalogue-item-ids) "catalogue item not specified")
+  (let [items (map (fn [id]
+                     (let [item (get-catalogue-item id)]
+                       (assert item (str "catalogue item " id " not found"))
+                       item))
+                   catalogue-item-ids)
+        _ (assert (= 1 (count (distinct (mapv :formid items)))) "catalogue items did not have the same form")
+        _ (assert (= 1 (count (distinct (mapv :wfid items)))) "catalogue items did not have the same workflow")
```
Comment (opqdonut) [inline src/clj/rems/application/commands.clj]: these could be errors output by the create command btw

File: src/clj/rems/db/applications.clj [hunk 3]
```diff
@@ -37,65 +37,28 @@
...
           last (apply max (cons 0 (map (comp read-string :suffix) all)))
           new (str (inc last))]
       (db/add-external-id! {:prefix prefix :suffix new})
       {:prefix prefix :suffix new})))
 
-(defn format-external-id [{:keys [prefix suffix]}]
+(defn- format-external-id [{:keys [prefix suffix]}]
   (str prefix "/" suffix))
 
 (defn application-external-id! [time]
```
Comment (opqdonut) [inline src/clj/rems/db/applications.clj]: private?

File: src/clj/rems/db/applications.clj [hunk 4]
```diff
@@ -109,18 +72,18 @@
    :secure-token secure-token
    :get-catalogue-item catalogue/get-localized-catalogue-item
    :get-catalogue-item-licenses get-catalogue-item-licenses
-   :application-created-event application-created-event})
+   :get-workflow workflow/get-workflow
+   :allocate-application-ids! allocate-application-ids!})
```
Comment (opqdonut) [inline src/clj/rems/db/applications.clj]: should this map be called command-injections?

File: src/clj/rems/db/applications.clj (hunk 3 above)
Comment (luontola) [reply src/clj/rems/db/applications.clj]: It's tested directly from `rems.db.test-applications/test-application-external-id!`

Comment (opqdonut) [reply src/clj/rems/db/applications.clj]: It's just kinda weird that allocate-application-ids! is private and this isn't. not a big thing but you could consider making this private and stealing it with `@#'` in the test.

**Pull request merged**